RDD not compiling

The language is not compiling on the latest Arch Linux.

System information
```
g++ (GCC) 14.2.1 20250207
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

clang version 19.1.7
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

<details>

<summary>Compile error</summary>

```
[ 50%] Building CXX object CMakeFiles/rdd.dir/main.cpp.o
In file included from /home/linux123123/Github/rdd/src/main.cpp:8:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘std::optional<Token> SyntaxAnalyzer::TryConsume(T) [with T = Tokens::NUMBER]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:29:43:   required from here
   29 |         if (const auto intLit = TryConsume(Tokens::NUMBER())) {
      |                                 ~~~~~~~~~~^~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:460:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::NUMBER’)
  460 |         if (Peek().has_value() && Peek().value().type == type)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘std::optional<Token> SyntaxAnalyzer::TryConsume(T) [with T = Tokens::STRING]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:39:47:   required from here
   39 |          if (const auto stringLit = TryConsume(Tokens::STRING())) {
      |                                     ~~~~~~~~~~^~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:460:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::STRING’)
  460 |         if (Peek().has_value() && Peek().value().type == type)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘std::optional<Token> SyntaxAnalyzer::TryConsume(T) [with T = Tokens::IDENTIFIER]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:49:42:   required from here
   49 |         if (const auto ident = TryConsume(Tokens::IDENTIFIER())) {
      |                                ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:460:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::IDENTIFIER’)
  460 |         if (Peek().has_value() && Peek().value().type == type)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘std::optional<Token> SyntaxAnalyzer::TryConsume(T, const std::string&) [with T = Tokens::DELIMITER; std::string = std::__cxx11::basic_string<char>]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:59:22:   required from here
   59 |         if(TryConsume(Tokens::DELIMITER(), "("))
      |            ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:453:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::DELIMITER’)
  453 |         if (Peek().has_value() && Peek().value().type == type && Peek().value().value == val)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘Token SyntaxAnalyzer::TryConsumeErr(T, const std::string&) [with T = Tokens::DELIMITER; std::string = std::__cxx11::basic_string<char>]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:62:30:   required from here
   62 |                 TryConsumeErr(Tokens::DELIMITER(), ")");
      |                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:444:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::DELIMITER’)
  444 |         if (Peek().has_value() && Peek().value().type == type && Peek().value().value == val)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘std::optional<Token> SyntaxAnalyzer::TryConsume(T, const std::string&) [with T = Tokens::KEYWORD; std::string = std::__cxx11::basic_string<char>]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:145:23:   required from here
  145 |         if (TryConsume(Tokens::KEYWORD(), "else if"))
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:453:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::KEYWORD’)
  453 |         if (Peek().has_value() && Peek().value().type == type && Peek().value().value == val)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘std::optional<Token> SyntaxAnalyzer::TryConsume(T, const std::string&) [with T = Tokens::OPERATOR; std::string = std::__cxx11::basic_string<char>]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:190:27:   required from here
  190 |             if (TryConsume(Tokens::OPERATOR(), ":")){ // Declaration
      |                 ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:453:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::OPERATOR’)
  453 |         if (Peek().has_value() && Peek().value().type == type && Peek().value().value == val)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘Token SyntaxAnalyzer::TryConsumeErr(T) [with T = Tokens::TYPENAME]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:191:49:   required from here
  191 |                 const Token type = TryConsumeErr(Tokens::TYPENAME());
      |                                    ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:435:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::TYPENAME’)
  435 |         if (Peek().has_value() && Peek().value().type == type)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘bool SyntaxAnalyzer::TryPeek(T, const std::string&, int) [with T = Tokens::DELIMITER; std::string = std::__cxx11::basic_string<char>]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:206:33:   required from here
  206 |                 else if (TryPeek(Tokens::DELIMITER(), ";")){  // Declaration without assignment
      |                          ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:430:70: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::DELIMITER’)
  430 |         return Peek(offset).has_value() && Peek(offset).value().type == type && Peek(offset).value().value == val;
      |                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘Token SyntaxAnalyzer::TryConsumeErr(T, const std::string&) [with T = Tokens::KEYWORD; std::string = std::__cxx11::basic_string<char>]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:288:30:   required from here
  288 |                 TryConsumeErr(Tokens::KEYWORD(), "to");
      |                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:444:55: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::KEYWORD’)
  444 |         if (Peek().has_value() && Peek().value().type == type && Peek().value().value == val)
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘bool SyntaxAnalyzer::TryPeek(T, int) [with T = Tokens::IDENTIFIER]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:317:25:   required from here
  317 |         else if (TryPeek(Tokens::IDENTIFIER())) {
      |                  ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:425:70: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::IDENTIFIER’)
  425 |         return Peek(offset).has_value() && Peek(offset).value().type == type;
      |                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp: In instantiation of ‘bool SyntaxAnalyzer::TryPeek(T, const std::string&, int) [with T = Tokens::OPERATOR; std::string = std::__cxx11::basic_string<char>]’:
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:318:24:   required from here
  318 |             if (TryPeek(Tokens::OPERATOR(), ":", 1) || TryPeek(Tokens::OPERATOR(), "=", 1)) {
      |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/linux123123/Github/rdd/src/SyntaxAnalyzer.hpp:430:70: error: no match for ‘operator==’ (operand types are ‘Tokens::Token*’ and ‘Tokens::OPERATOR’)
  430 |         return Peek(offset).has_value() && Peek(offset).value().type == type && Peek(offset).value().value == val;
      |                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
make[2]: *** [CMakeFiles/rdd.dir/build.make:79: CMakeFiles/rdd.dir/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:87: CMakeFiles/rdd.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```
</details>

--- a/src/Token.hpp
+++ b/src/Token.hpp
@@ -250,6 +250,15 @@ namespace Tokens {
         {new IDENTIFIER(), IDENTIFIER().regex()},
         {new OPERATOR(), OPERATOR().regex()},
     };
+
+    // Add a generic operator== to compare Tokens::Token* with specific token types
+    inline bool operator==(const Token* token, const Token& type) {
+        return typeid(*token) == typeid(type);
+    }
+
+    inline bool operator!=(const Token* token, const Token& type) {
+        return !(token == type);
+    }
 }
 
 class Token {

--- a/src/Token.hpp
+++ b/src/Token.hpp
@@ -257,7 +257,7 @@ namespace Tokens {
     }
 
     inline bool operator!=(const Token* token, const Token& type) {
-        return !(token == type);
+        return typeid(*token) != typeid(type);
     }
 }
 

